test/notification.go

```diff
@@ -0,0 +1,24 @@
+package test
+
+import (
+	"testing"
+
+	"fyne.io/fyne"
+	"github.com/stretchr/testify/assert"
+)
+
+// ExpectNotification allows an app developer to assert that a notification was sent.
+// After the content of f has executed this utility will check that the specified notification was sent.
+func ExpectNotification(t *testing.T, n *fyne.Notification, f func(a fyne.App)) {
+	a := NewApp().(*testApp)
```

Comment:
Why not simply use the current app? When using the `test` package there should be one.

Reply:
*test.App is not the app type, it's *test.testApp - i.e. not exported. Not sure I see the benefit in creating a new type just for this callback.
In this instance it is critical that the CurrentApp() is a *test.testApp, so that we can catch the notification.

Inside the test package you could set a different current app for whatever purposes, so creatting a clean one (as we do in some of our tests) seems like the only approach.

Reply:
Ah, I had forgotten testapp is not exported. Nevertheless I don't think you should create a new testapp here - if any of the dev's test following this call depends on something they've setup in their testapp it will fail in a non-obvious way.

I think it would be better to follow @toaster's suggestion, with an type cast:
```suggestion
	var ta *testApp
	if fyne.CurrentApp() == nil {
		ta = NewApp()
	} else {
		a, ok := fyne.CurrentApp().(*testApp)
		if ok {
			ta = a
		} else {
			t.Fail("Must use test app for notification assertion")
			return
		}
	}
	// Use ta
```

Reply:
I don't really want to fail a test due to internal implementation which could have completed successfully...

Reply:
Fair enough, instead create a new testapp if you can't use the current one, but at least log that this has happened:
```
} else {
    fyne.LogError("Cannot use current app for notification assertion, creating new test app", nil)
    ta = NewApp()
}
```

Reply:
If the concern is the side-effects of setting an app directly then I guess I agree. For that reason - and for the "app.New(); defer app.New()" pattern that is niggling me - I propose something like "UsingApp(fyne.App, func())" utility.

Then this would be:
```
UsingApp(NewApp(), func() {
    ... do stuff to test notification ...
})

func UsingApp(a fyne.App, fn func()) {
    oldApp := fyne.CurrentApp()
    fyne.SetCurrentApp(a)
    fn()
    fyne.SetCurrentApp(oldApp)
}
```

Reply:
Hypothetically, the dev wants to write a test that something triggers a notification only once. They make notification1, call some code that sends it, and finally assert it is sent. Then they make notification2, call some code that should not send it, and finally asserts that the last notification sent to the user was notification1, but instead get "Expected notification1, got nil" <- this is not obvious to the dev at all

Reply:
The method is now named `AssertNotificationSent` - I think it would be expected that this fails due to nil not being the same as notification2.
I'm not sure where the "last notification" comes from - that's not provided by this supporting API.

Reply:
I just don't think this should create an app, it should either accept App and type cast to testApp, or use fyne.CurrentApp() and throw nil pointer if the dev didn't set one up. That would be less surprising than an Assert creating apps behind the scenes.

Reply:
I doon't understand the concern of creating a new app. This test requires that a testApp be the current app. And to avoid corner cases we also need it to be clean (if the previous notification is the the same as a new expected notification that does not arrive it could pass incorrectly...